### ISSUE TITLE
Yolo Dataset Split Overlap Issue in Test Cases

### ISSUE THREAD

**Issue by S0nFra** [freedreamer82/saltup#13]
The test `test_split_dataset()` in `saltup/test/ai/object_detection/dataset/test_yolo_dataset.py` frequently fails with the following error:

```bash
============================= test session starts ==============================
platform linux -- Python 3.12.2, pytest-8.3.4, pluggy-1.5.0
rootdir: /home/fsonnessa/Documents/GitHub/saltup
configfile: pyproject.toml
collected 10 items

test/ai/object_detection/dataset/test_yolo_dataset.py ......F...         [100%]

=================================== FAILURES ===================================
______________________ TestYOLODarknet.test_split_dataset ______________________

self = <object_detection.dataset.test_yolo_dataset.TestYOLODarknet object at 0x745b260167b0>
sample_dataset = (PosixPath('/tmp/pytest-of-fsonnessa/pytest-2/test_split_dataset0/dataset'), {'images': {'test': '/tmp/pytest-of-fsonn...it_dataset0/dataset/labels/train', 'val': '/tmp/pytest-of-fsonnessa/pytest-2/test_split_dataset0/dataset/labels/val'}})

    def test_split_dataset(self, sample_dataset):
        """Test splitting dataset while maintaining class distribution."""
        root_dir, dirs = sample_dataset
    
        # Get class distribution
        class_to_images = _image_per_class_id(
            str(dirs['labels']['train']),
            str(dirs['images']['train'])
        )
    
        # Test splitting
        train, val, test = split_dataset(
            class_to_images,
            split_ratio=0.8,
            split_val_ratio=0.5
        )
    
        # Verify splits
        total_images = sum(len(imgs) for imgs in class_to_images.values())
        total_split = len(train) + len(val) + len(test)
        assert total_split <= total_images
    
        # Check no overlap between splits
        train_set = set(str(x) for x in train)
        val_set = set(str(x) for x in val)
        test_set = set(str(x) for x in test)
        assert not (train_set & val_set)
>       assert not (train_set & test_set)
E       assert not ({"('/tmp/pytest-of-fsonnessa/pytest-2/test_split_dataset0/dataset/images/train/img1.jpg', '/tmp/pytest-of-fsonnessa/py...ataset/images/train/img4.jpg', '/tmp/pytest-of-fsonnessa/pytest-2/test_split_dataset0/dataset/labels/train/img4.txt')"} & {"('/tmp/pytest-of-fsonnessa/pytest-2/test_split_dataset0/dataset/images/train/img1.jpg', '/tmp/pytest-of-fsonnessa/pytest-2/test_split_dataset0/dataset/labels/train/img1.txt')"})

test/ai/object_detection/dataset/test_yolo_dataset.py:198: AssertionError
=========================== short test summary info ============================
FAILED test/ai/object_detection/dataset/test_yolo_dataset.py::TestYOLODarknet::test_split_dataset
========================= 1 failed, 9 passed in 0.11s ==========================
```

## Issue Description

The test failure indicates an overlap between training and test sets (and potentially between training and validation sets) depending on how `random.shuffle()` in `saltup.ai.object_detection.dataset.yolo_darknet.split_dataset` mixes the samples.

The test uses a sample dataset of 4 images. The objects in these images are distributed across three classes as follows:

```python
# <class_id>: <images_with_object_in_class>
{
    0: ['img1','img3'],
    1: ['img1','img3','img2'],
    2: ['img4','img2'],
}
```

## Analysis

I manually traced the split function execution with these parameters:
```python
split_ratio = 0.8
split_val_ratio = 0.5
```

Step-by-step execution:
```
- Step 1
limit = 2
train_count = 1
train_set = {img1}
val_test_set = {img3}

- Step 2
limit = 3
train_count = 2
train_set = {img1, img3}
val_test_set = {img3, img2}

- Step 3
limit = 2
train_count = 1
train_set = {img1, img3, img4}
val_test_set = {img3, img2}

- Final Result -
mid_point = 1
train_set = {img1, img3, img4}
val_set = {img3}
test_set = {img2}
```

The issue appears to stem from the way sets are populated in `split_dataset()`. The problematic code section is:

```python
# yolo_darknet.py:461
train_set, val_test_set = set(), set()

for class_id, images in class_to_images.items():
    random.shuffle(images)
    limit = min(max_images_per_class, len(images)) if max_images_per_class else len(images)

    # Split into training and remaining sets
    train_count = int(limit * split_ratio)
    train_set.update(images[:train_count])
    val_test_set.update(images[train_count:limit])
```

## Impact

While this overlap is particularly noticeable in small test dataset of four images, the impact on larger datasets needs investigation. Although our production model shows good performance, this issue should be addressed to ensure proper dataset separation.

## Next Steps

1. I'm working on a potential solution to prevent dataset overlap
2. Request for other team members to verify this issue and confirm my findings
3. Once verified, we should implement a fix that ensures strict separation between training, validation, and test sets

Would appreciate any feedback or additional insights from the team.
